src/duty/duty.service.ts

```diff
@@ -0,0 +1,70 @@
+import { LOGGER_PROVIDER } from '@lido-nestjs/logger';
+import { Inject, Injectable, LoggerService } from '@nestjs/common';
+
+import { ConfigService } from 'common/config';
+import { ConsensusProviderService } from 'common/eth-providers';
+import { BlockCacheService } from 'common/eth-providers/consensus-provider/block-cache';
+import { bigintRange } from 'common/functions/range';
+import { PrometheusService, TrackTask } from 'common/prometheus';
+import { ClickhouseService } from 'storage';
+
+import { AttestationService } from './attestation';
+import { ProposeService } from './propose';
+import { StateService } from './state';
+import { SummaryService } from './summary';
+import { SyncService } from './sync';
+
+@Injectable()
+export class DutyService {
+  public constructor(
+    @Inject(LOGGER_PROVIDER) protected readonly logger: LoggerService,
+    protected readonly config: ConfigService,
+    protected readonly prometheus: PrometheusService,
+    protected readonly clClient: ConsensusProviderService,
+    protected readonly blockCacheService: BlockCacheService,
+
+    protected readonly state: StateService,
+    protected readonly attestation: AttestationService,
+    protected readonly propose: ProposeService,
+    protected readonly sync: SyncService,
+    protected readonly summary: SummaryService,
+    protected readonly storage: ClickhouseService,
+  ) {}
+
+  public async checkAndWrite(epoch: bigint, stateSlot: bigint): Promise<any> {
+    await Promise.all([this.prefetch(epoch), this.checkAll(epoch, stateSlot)]);
```

Comment:
Shouldn't the **prefetch** task be done before checks?

Reply:
`prefetch` will be done before main checks because `duty by state` requests are heavy and while we wait for their responses we fetch blocks and headers. If for some reason `prefetch` task will be slower than `duty by state` requests, blocks and headers will be fetched inside tasks of checks

Reply:
Leave this comment in the code, it's important.

Reply:
done